Renewed PULL for merging the branches

CKehl opened the pull request:
Dear osgAndroid team,

we've been mailing each other around half a year ago. Thing is that back then, I made one, big commit and it's hard to cluster/de-cluster all the changes.

On the commit after that (c178f04b2b86ec889917b3d17d2d7e9f90a41d59) I reset back to your old main branch (last commit I had from y'all) (commit: e49ad0dc74567c6052b32e0f581c7dbd42823713; Fixed Jordi Torres name in the readme). So, from commit c178f04b2b86ec889917b3d17d2d7e9f90a41d59 onwards I commit each change piece-by-piece so you can merge that calmly and easily get back to me if there are errors in the meanwhile.

Kind Regards,

Christian

Commit:
Added modification, including:
- osg Image Reader/Writer
- osg Node Reader/Writer
- Off-Screen Rendering (using Pixel Buffer Objects)
- wrapper for Vec2/Vec3/Vec4 and their Arrays
- extended wrapper for the Matrix
- basic classes to construct a renderable object Mesh (Geode, Geometry, ...)
- wrapped native functions for texture projection and Raycasting intersection

Commit:
changed Android.mk to the <path-to-osg-gles1-sdk> variant

Commit:
added license and contribution information

Commit:
added new compilation files that include 3rdParty libraries and that works for GLES1 and GLES2

Commit:
added ignore rule for temporary gedit backup files (Ubuntu text editor).

Commit:
Changes to be committed:
	modified:   jni/JNIosg.cpp

expanded wrapped C++ functionality (functions) in the osg Core domain of the following classes:
Node
- setTexture
- setPointSize
- setLineWidth

Group
- addChild
- removeChild
- getNumberOfChildren
- getChild

Camera
- setClearColorVec
- getClearColorVec
- getViewMatrix

Vec2 (new)
- dispose, create
- x,y
- set, setX, setY
- length, length2
- div, dotProduct, scalarProduct, sum, sub
- normalize
- negation

Matrix
- inverse, transpose
- scale, mult
- get, getTranslation, getRotation
- [pre/post]MultVec[3/4]

Image
- getRed/Green/Blue/Alpha
- s,t

Commit:
Changes to be committed:
	modified:   src/org/openscenegraph/osg/core/Camera.java
	modified:   src/org/openscenegraph/osg/core/Group.java
	modified:   src/org/openscenegraph/osg/core/Image.java
	modified:   src/org/openscenegraph/osg/core/Matrix.java
	modified:   src/org/openscenegraph/osg/core/Node.java

    expanded Java functionality (functions), which C++ function calls
    are previously added, in the osg Core domain of the following classes:
    Node
    - setTexture
    - setPointSize
    - setLineWidth

    Group
    - addChild
    - removeChild
    - getNumberOfChildren
    - getChild

    Camera
    - setClearColorVec
    - getClearColorVec
    - getViewMatrix

    Vec2 (new)
    - dispose, create
    - x,y
    - set, setX, setY
    - length, length2
    - div, dotProduct, scalarProduct, sum, sub
    - normalize
    - negation

    Matrix
    - inverse, transpose
    - scale, mult
    - get, getTranslation, getRotation
    - [pre/post]MultVec[3/4]

    Image
    - getRed/Green/Blue/Alpha
    - s,t

Commit:
Changes to be committed:
	new file:   src/org/openscenegraph/osg/core/Array.java
	new file:   src/org/openscenegraph/osg/core/DrawArrays.java
	new file:   src/org/openscenegraph/osg/core/DrawElementsUInt.java
	new file:   src/org/openscenegraph/osg/core/Drawable.java
	new file:   src/org/openscenegraph/osg/core/Geode.java
	new file:   src/org/openscenegraph/osg/core/Geometry.java
	new file:   src/org/openscenegraph/osg/core/Object.java
	new file:   src/org/openscenegraph/osg/core/PrimitiveSet.java
	new file:   src/org/openscenegraph/osg/core/PrimitiveSetList.java
	new file:   src/org/openscenegraph/osg/core/Texture.java
	new file:   src/org/openscenegraph/osg/core/Texture2D.java
	new file:   src/org/openscenegraph/osg/core/Vec2.java
	new file:   src/org/openscenegraph/osg/core/Vec2Array.java
	modified:   src/org/openscenegraph/osg/core/Vec3.java
	new file:   src/org/openscenegraph/osg/core/Vec3Array.java
	modified:   src/org/openscenegraph/osg/core/Vec4.java
	new file:   src/org/openscenegraph/osg/core/Vec4Array.java

added a large collection of Java classes to interface via Android SDK;
because the osg::Object is now wrapped, a minor adaptation had to be
done for Vec3 and Vec4 in the equals() method

Commit:
Changes to be committed:
	modified:   jni/JNIosgViewer.cpp
	new file:   jni/screenview.cpp
	new file:   jni/screenview.h

Added newly wrapped C++ functions to allow for off-screen rendering
on Android. The screenview class is adapted from the public OSG example.
The viewer, formerly just being JNIviewer, is now separated into Viewer
and OffScreenViewer, which both inherit from ViewerBase.

Commit:
Changes to be committed:
	modified:   src/org/openscenegraph/osg/viewer/OSGRenderer.java
	new file:   src/org/openscenegraph/osg/viewer/OffScreenViewer.java
	modified:   src/org/openscenegraph/osg/viewer/Viewer.java
	new file:   src/org/openscenegraph/osg/viewer/ViewerBase.java

Added newly JNI functions and classes to allow for off-screen rendering
on Android.
The viewer, formerly just being JNIviewer, is now separated into Viewer
and OffScreenViewer, which both inherit from ViewerBase.

NOTE: just ran an ndk build - errors came up as the normally distributed
OSG version apparently doesn't have a RefVec2 (just the way it provides
one for Vec3 and Vec4). I know I added it locally to an OSG build (only
thing I changed is defining the RefVec2) and submitted it to osg-submissions
mailing list. please check in the osg trunk of 3.4 or later if it's there
for you - OR: add it as external definition is the JNI headers here. Don't
know what's more handy.

Commit:
Changes to be committed:
	modified:   jni/Android.mk
	modified:   jni/JNIUtils.h
	modified:   jni/JNIosg.cpp

BUG SOLVED:
Found the RefVec2 error - I forgot to add it to the JNIUtils (done now).
Please check your osg installation if ReferenceType<osg::Vec2> is defined.
Also added the screenview.h/cpp to Android.mk to make the OffScreenViewer working.
Removed an absurdly large test function for vertex coloring so to remove the request
for an interpolation function (added soon).

ndk-build compiles now again without errors.

Commit:
Changes to be committed:
	modified:   jni/JNIUtils.h

Added Interpolate() function

Commit:
Changes to be committed:
	modified:   jni/JNIosgDB.cpp

Added support (in the C++ wrappers) for using the plugins automatically.
Also provided functions for ReadImage, WriteNode and WriteImage.

Commit:
Changes to be committed:
	modified:   src/org/openscenegraph/osg/db/ReadFile.java
	new file:   src/org/openscenegraph/osg/db/WriteFile.java

Added Java-callable classes (or functions) to read or write a node or
an image.

Commit:
Changes to be committed:
	modified:   jni/Android.mk
	modified:   jni/JNIosgDB.cpp

BUG SOLVED:
changed the log message of osgDB to avoid conflicts occurring from
the recent addition. Also reverted the Android.mk OSG_HEAD definition
to the <path-to....> one.

Commit:
Changes to be committed:
	new file:   LICENSE
	new file:   Readme.md

 Changes not staged for commit:
	deleted:    .classpath
	deleted:    .project

Added LICENSE and Readme file as given in the main trunk,
and added my name to the contributors.

MERGE COMPLETE

Commit:
Merge branch 'master' of https://github.com/CKehl/osgAndroid

Conflicts:
	org.openscenegraph.android/jni/Android.mk
	org.openscenegraph.android/jni/JNIosg.cpp
	org.openscenegraph.android/jni/JNIosgViewer.cpp

 All conflicts fixed but you are still merging.

 Changes to be committed:
	modified:   org.openscenegraph.android/Readme.md
	modified:   org.openscenegraph.android/jni/Android.mk
	modified:   org.openscenegraph.android/jni/Application.mk
	modified:   org.openscenegraph.android/jni/JNIUtils.h
	modified:   org.openscenegraph.android/jni/JNIosg.cpp
	modified:   org.openscenegraph.android/jni/JNIosgUtil.cpp
	modified:   org.openscenegraph.android/jni/JNIosgViewer.cpp
	modified:   org.openscenegraph.android/src/org/openscenegraph/osg/util/GeometryUtils.java
	modified:   org.openscenegraph.android/src/org/openscenegraph/osg/viewer/Viewer.java

 Changes not staged for commit:
	deleted:    org.openscenegraph.android/.classpath
	deleted:    org.openscenegraph.android/.project

Merging old head with reverted new history - hope that works on remote

Commit:
Changes to be committed:
	modified:   Readme.md
	deleted:    org.openscenegraph.android/LICENSE
	deleted:    org.openscenegraph.android/Readme.md

removed confusion on doubled LICENSE and Readme.md

jtorresfabra commented:
Hi @CKehl, 

Thank you very much for taking the time to redo your PR, we know it has been a hard effort. We will be reviewing it in the next few days.  

Cheers!

jtorresfabra commented on org.openscenegraph.android/jni/Android.mk:
```diff
@@ -57,45 +66,111 @@ OSG_LDLIBS := \
 
+### Static preparation
+OSG_HEAD:=<path-to-osg-gles1-sdk> 
+#OSG_HEAD:=/media/christian/DATA/android-osg-sdk/gles1
```
Remove comments with fixed paths

CKehl replied:
will be done

jtorresfabra commented on org.openscenegraph.android/jni/Android.mk:
```diff
@@ -57,45 +66,111 @@ OSG_LDLIBS := \
-OSG_SDK_LIB_PATH := $(OSG_SDK)/lib
-OSG_SDK_PLUGIN_PATH := $(OSG_SDK_LIB_PATH)/osgPlugins-3.3.8
+OSG_HEAD:=<path-to-osg-gles2-sdk>
+#OSG_HEAD:=/media/christian/DATA/android-osg-sdk/gles2
```
Same as above

jtorresfabra commented on org.openscenegraph.android/jni/Android.mk:
```diff
@@ -57,45 +66,111 @@ OSG_LDLIBS := \
-OSG_SDK_LIB_PATH := $(OSG_SDK)/lib
-OSG_SDK_PLUGIN_PATH := $(OSG_SDK_LIB_PATH)/osgPlugins-3.3.8
+OSG_HEAD:=<path-to-osg-gles1-sdk>
+#OSG_HEAD:=/media/christian/DATA/android-osg-sdk/gles1
```
Same as above

jtorresfabra commented:
@CKehl  as a general comment, we were waiting different PR's for different features, not different commits. this way it is easy to use the github interface searching for changes in pull requests. But ok, we will try to review it as is.

jtorresfabra commented on org.openscenegraph.android/jni/JNIUtils.h:
```diff
@@ -37,6 +38,51 @@ class ReferencedType: public T, public osg::Referenced
 typedef ReferencedType<osg::Vec4> RefVec4;
 typedef ReferencedType<osg::Quat> RefQuat;
+
+inline osg::Vec4 Interpolate(osg::Vec2 _texcoord, osg::Image* _img)
```
shouldn't it be "const osg::Vec2 &_texcoord"? Also we tend to use underscore for member class variables, not for parameters.

CKehl replied:
I agree Jordi: for efficiency reasons, it can be "const osg::Vec2 &texcoord" - I just needed this function quickly, so I wasn't that picky with the reference. The underscoring: noted and will be changed.

CKehl commented:
Op 06-09-16 om 15:07 schreef Jordi Torres:

> @CKehl https://github.com/CKehl as a general comment, we were 
> waiting different PR's for different features, not different commits. 
> this way it is easy to use the github interface searching for changes 
> in pull requests. But ok, we will try to review it as is.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> https://github.com/miragetech/osgAndroid/pull/2#issuecomment-244944083, 
> or mute the thread 
> https://github.com/notifications/unsubscribe-auth/AFvds2HdtVJ6crTgT6Q7b8SEmf1cHozZks5qnWWWgaJpZM4Jxzpn.
> 
> Okay, I didn't know how it works (new to it). It would be nice if you 
> can try review it with the different commit. Otherwise, I need to create 
> a new branch from the main, roll back the whole main and redo all the 
> changes again with separate PRs.

How do you go now about the commenting ? should I just erase the 
fixed-path comments in the future ?

jtorresfabra commented:
Hi @CKehl, 

The idea is that you fix yourself your branch, it automatically appears fixed in the PR.

Thanks.

jtorresfabra commented on org.openscenegraph.android/jni/JNIosg.cpp:
```diff
@@ -139,6 +178,931 @@ JNIEXPORT jint JNICALL Java_org_openscenegraph_osg_core_Group_nativeGetNumChildr
+}
+
+/*
+JNIEXPORT jlong JNICALL Java_org_openscenegraph_osg_core_PrimitiveSet_nativeCreatePrimitiveSet(JNIEnv *, jclass, jint type, jint mode)
```
Better if we do not leave dead code, if it does not work/serve anymore is better to remove it.

CKehl replied:
okay.

CKehl commented:
Op 06-09-16 om 15:19 schreef Jordi Torres:

> Hi @CKehl https://github.com/CKehl,
> 
> The idea is that you fix yourself your branch, it automatically 
> appears fixed in the PR.
> 
> Thanks.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> https://github.com/miragetech/osgAndroid/pull/2#issuecomment-244947195, 
> or mute the thread 
> https://github.com/notifications/unsubscribe-auth/AFvds9fbp9RetIMCFaKrMrN-aiy_-q0yks5qnWhggaJpZM4Jxzpn.
> 
> Hi,

well, that was not the way I meant it. I guess I understand you 
correctly that you want to have a single pull request for each tested 
and final version of a feature of my branch. What you now have is one PR 
with different commits per feature, tested and working. Sorry for that.

jtorresfabra commented on org.openscenegraph.android/jni/screenview.cpp:
```diff
@@ -0,0 +1,389 @@
+    getSize(_gc, width, height);
+    if (width!=_width || _height!=height)
+    {
+        std::cout<<"   Window resized "<<width<<", "<<height<<std::endl;
```
rather to  use osg::notify mechanism

CKehl replied:
okay, will be adapted

jtorresfabra commented on org.openscenegraph.android/jni/screenview.cpp:
```diff
@@ -0,0 +1,389 @@
+
+    updateTimings(tick_start, tick_afterReadPixels, tick_afterMemCpy, image->getTotalSizeInBytes());
+
+    if (!_fileName.empty())
```
Dead code.

CKehl replied:
I just copied the plain example for offscreen rendering from the web, which worked on the desktop and which I wrapped as-is. I can have a look on it again, but it may also be helpful if you could point me to specifics in the function that are 'dead' by now.

jtorresfabra replied:
I mean: 
`if (!_fileName.empty())
  {
￼         // osgDB::writeImageFile(*image, _fileName);
￼  }`

This code does nothing and should be removed, maybe the example is also wrong, I don't recall.

CKehl replied:
Now I see and agree. The original version was targeted towards storing the off-screen buffer image to disc, but that one I took out in the header (we can do that via Java). So yes, I will take out that passage. Thanks for clearifying!

jtorresfabra commented:
Hey @CKehl,

https://help.github.com/articles/about-pull-requests/

So "Once you've created a pull request, you can push commits from your
topic branch to add them to your existing pull request. These commits will
appear in chronological order within your pull request and the changes will
be visible in the "Files changed" tab."

So the only thing you need is to keep adding commits with the fixes to your
PR. It should be enough.

Does it make sense for you?

2016-09-06 15:27 GMT+02:00 CKehl notifications@github.com:

> Op 06-09-16 om 15:19 schreef Jordi Torres:
> 
> > Hi @CKehl https://github.com/CKehl,
> > 
> > The idea is that you fix yourself your branch, it automatically
> > appears fixed in the PR.
> > 
> > Thanks.
> > 
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > https://github.com/miragetech/osgAndroid/pull/2#issuecomment-244947195,
> > 
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-
> > auth/AFvds9fbp9RetIMCFaKrMrN-aiy_-q0yks5qnWhggaJpZM4Jxzpn>.
> > 
> > Hi,
> 
> well, that was not the way I meant it. I guess I understand you
> correctly that you want to have a single pull request for each tested
> and final version of a feature of my branch. What you now have is one PR
> with different commits per feature, tested and working. Sorry for that.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/miragetech/osgAndroid/pull/2#issuecomment-244949481,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/ABs8RMTW2FuTjXK7vnYbqevuUv0CnP4Cks5qnWpWgaJpZM4Jxzpn
> .

## 

Jordi Torres

jtorresfabra commented on org.openscenegraph.android/jni/screenview.h:
```diff
@@ -0,0 +1,273 @@
+
+                getSize(gc, _width, _height);
+
+                std::cout<<"Window size "<<_width<<", "<<_height<<std::endl;
```
rather use osg::notify.

CKehl replied:
already noted.

CKehl commented:
Well seem's at least I have a little read ahead of me :-) But in 
essence: yes, got that.

Op 06-09-16 om 15:37 schreef Jordi Torres:

> Hey @CKehl,
> 
> https://help.github.com/articles/about-pull-requests/
> 
> So "Once you've created a pull request, you can push commits from your
> topic branch to add them to your existing pull request. These commits will
> appear in chronological order within your pull request and the changes 
> will
> be visible in the "Files changed" tab."
> 
> So the only thing you need is to keep adding commits with the fixes to 
> your
> PR. It should be enough.
> 
> Does it makes sense for you?
> 
> 2016-09-06 15:27 GMT+02:00 CKehl notifications@github.com:
> 
> > Op 06-09-16 om 15:19 schreef Jordi Torres:
> > 
> > > Hi @CKehl https://github.com/CKehl,
> > > 
> > > The idea is that you fix yourself your branch, it automatically
> > > appears fixed in the PR.
> > > 
> > > Thanks.
> > > 
> > > —
> > > You are receiving this because you were mentioned.
> > > Reply to this email directly, view it on GitHub
> > > 
> > > https://github.com/miragetech/osgAndroid/pull/2#issuecomment-244947195,
> > > 
> > > or mute the thread
> > > <https://github.com/notifications/unsubscribe-
> > > auth/AFvds9fbp9RetIMCFaKrMrN-aiy_-q0yks5qnWhggaJpZM4Jxzpn>.
> > > 
> > > Hi,
> > 
> > well, that was not the way I meant it. I guess I understand you
> > correctly that you want to have a single pull request for each tested
> > and final version of a feature of my branch. What you now have is one PR
> > with different commits per feature, tested and working. Sorry for that.
> > 
> > —
> > You are receiving this because you commented.
> > Reply to this email directly, view it on GitHub
> > 
> > https://github.com/miragetech/osgAndroid/pull/2#issuecomment-244949481,
> > or mute the thread
> > 
> > https://github.com/notifications/unsubscribe-auth/ABs8RMTW2FuTjXK7vnYbqevuUv0CnP4Cks5qnWpWgaJpZM4Jxzpn
> > .
> 
> ## 
> 
> Jordi Torres
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> https://github.com/miragetech/osgAndroid/pull/2#issuecomment-244952130, 
> or mute the thread 
> https://github.com/notifications/unsubscribe-auth/AFvds438gtxuSsS7bQpur4xHYTN8tOMLks5qnWyHgaJpZM4Jxzpn.

rafagaitan commented on org.openscenegraph.android/jni/JNIosg.cpp:
```diff
@@ -93,14 +113,33 @@ JNIEXPORT void JNICALL Java_org_openscenegraph_osg_core_Node_nativeSetMode(JNIEn
+	if(node != NULL)
+		node->getOrCreateStateSet()->setAttributeAndModes(lw, osg::StateAttribute::ON);
+}
+
```
I'm wondering if it would be better to have StateSet abstraction with those methods (including the nativeSetMode), maybe, for now this is ok, but I will open a issue for tracking this change soon.

CKehl replied:
Rafa, I agree. But messing with the StateSet wrapping is something I wanted to avoid for now - especially as I usually rather work with geometry, hence more often work with the node concept.

rafagaitan commented on org.openscenegraph.android/jni/JNIosg.cpp:
```diff
@@ -139,6 +178,931 @@ JNIEXPORT jint JNICALL Java_org_openscenegraph_osg_core_Group_nativeGetNumChildr
+
+	int array_n = env->GetArrayLength(inputBuffer);
+	//printf("Array size: %i\n", array_n);
+	osg::Vec3Array* _array = new osg::Vec3Array();
```
We prefer local variables name without underscore, change "_array" to "array". We only use the underscore in class attributes.

CKehl replied:
mentioned by Jordi already. Will be changed by 14/09/2016.

rafagaitan commented on org.openscenegraph.android/jni/JNIosg.cpp:
```diff
@@ -139,6 +178,931 @@ JNIEXPORT jint JNICALL Java_org_openscenegraph_osg_core_Group_nativeGetNumChildr
+	osg::Geometry *g = reinterpret_cast<osg::Geometry*>(cptr);
+
+	int array_n = env->GetArrayLength(inputBuffer);
+	osg::Vec3Array* _array = new osg::Vec3Array();
```
Same as above

rafagaitan commented on org.openscenegraph.android/jni/JNIosg.cpp:
```diff
@@ -139,6 +178,931 @@ JNIEXPORT jint JNICALL Java_org_openscenegraph_osg_core_Group_nativeGetNumChildr
+	int array_n = env->GetArrayLength(inputBuffer);
+
+	//LOGI("Colour Array size: %i\n", array_n);
+	osg::Vec4Array* _array = new osg::Vec4Array();
```
Same as above

rafagaitan commented on org.openscenegraph.android/jni/JNIosg.cpp:
```diff
@@ -139,6 +178,931 @@ JNIEXPORT jint JNICALL Java_org_openscenegraph_osg_core_Group_nativeGetNumChildr
+		}
+	}
+
+	if(g!=NULL)
```
probably early checking at the beginning of the function
`if (g == NULL)
     return;
`
will avoid all computations if g is not valid

CKehl replied:
agreed. I will go over it again and adapt late terminations towards early terminations.

rafagaitan commented on org.openscenegraph.android/jni/JNIosg.cpp:
```diff
@@ -139,6 +178,931 @@ JNIEXPORT jint JNICALL Java_org_openscenegraph_osg_core_Group_nativeGetNumChildr
+	osg::Geometry *g = reinterpret_cast<osg::Geometry*>(cptr);
+
+	int array_n = env->GetArrayLength(inputBuffer);
+	osg::Vec2Array* _array = new osg::Vec2Array();
```
remove underscore

rafagaitan commented on org.openscenegraph.android/jni/JNIosg.cpp:
```diff
@@ -139,6 +178,931 @@ JNIEXPORT jint JNICALL Java_org_openscenegraph_osg_core_Group_nativeGetNumChildr
+{
+	osg::Drawable *g = reinterpret_cast<osg::Drawable *>(cptr);
+	return reinterpret_cast<jlong>(g);
+}
```
This function, from the JNI/Java point of view does not do anything, the casting/implementation must be done at the Java layer

CKehl replied:
when I see it now: yes, right. I'll move it to the Java part.

rafagaitan commented on org.openscenegraph.android/jni/JNIosg.cpp:
```diff
@@ -139,6 +178,931 @@ JNIEXPORT jint JNICALL Java_org_openscenegraph_osg_core_Group_nativeGetNumChildr
+{
+	osg::PrimitiveSet *g = reinterpret_cast<osg::PrimitiveSet *>(cptr);
+	return reinterpret_cast<jlong>(g);
+}
```
same as my other comment, asDrawable, asWhatever is a c++ smart way to avoid the dynamic_cast, in this case we only know about "jlong" pointers which are converted using reinterpret_cast, which forces the type, so.. better implement this at the Java level and remove it from here.

rafagaitan commented on org.openscenegraph.android/jni/JNIosgDB.cpp:
```diff
@@ -85,4 +95,40 @@ JNIEXPORT jlong JNICALL Java_org_openscenegraph_osg_db_ReadFile_nativeReadNodeFi
+
+JNIEXPORT jboolean JNICALL Java_org_openscenegraph_osg_db_WriteFile_nativeWriteNodeFile(JNIEnv* env, jclass, jstring filename, jlong node_cptr)
+{
+	#define LOG_TAG "org.openscenegraph.osg.db"
```
is this correct here?

CKehl replied:
hm - somewhat can remember that I had to do this somewhat messy define here - but I don't know why anymore. I'll try to remove it and see if it still works.

rafagaitan commented on org.openscenegraph.android/jni/JNIosgDB.cpp:
```diff
@@ -85,4 +95,40 @@ JNIEXPORT jlong JNICALL Java_org_openscenegraph_osg_db_ReadFile_nativeReadNodeFi
+
+JNIEXPORT jboolean JNICALL Java_org_openscenegraph_osg_db_WriteFile_nativeWriteImageFile(JNIEnv* env, jclass, jstring filename, jlong image_cptr)
+{
+	#define LOG_TAG "org.openscenegraph.osg.db"
```
same as above

rafagaitan commented on org.openscenegraph.android/jni/JNIosgUtil.cpp:
```diff
@@ -128,4 +180,284 @@ JNIEXPORT jlong JNICALL Java_org_openscenegraph_osg_util_GeometryUtils_nativeCre
+
+	osgDB::writeNodeFile(*_node, jstring2string(env,out_filepath));
+	return reinterpret_cast<jint>(colored_point);
+}
```
Could you explain the purpose of these two functions?
They seem as something you required for your application, I think they could be integrated but I want to understand the real purpose and maybe put them in the proper place.

CKehl replied:
The functions colour(s) a point/set of points using a given image texture that is connected with a camera pose facing the geometry (in world coordinate system). Thus, the texture coordinate parameterization established in the mid is just valid for a given camera position (which is expressed as a transformation matrix of the camera) the function determines the local image coordinate for the/each point and, if within the camera frame, colours the point accordingly.
To the parameters: trmat is the camera matrix, R is the intrinsic camera matrix (holding the frame extend in metres and the focal length), and Cg is the camera center vector.

CKehl replied:
wouldn't be a massive 'ouch' to remove it from the official trunk - but I need it in my application and my trunk. Perhaps I create then a special branch with that function for me.

Commit:
Changes to be committed:
	modified:   jni/Android.mk

Fixed removing comments from Android.mk

Commit:
stripped underscored variables and unnecessary comments from JNIosg.cpp

Commit:
Added early-fail termination via "if(var==NULL) return;" for array setter functions

Commit:
removed the as<SuchAndSuch> wrapper functions in JNIosg.cpp

Commit:
osg.Core change requests completed.
cleared the large TextureFromPose-functions.

Commit:
Fixed the in-function defined, which are not needed anymore as it appears

Commit:
modified:   JNIosgViewer.cpp

removed underscored variable names, unnecessary comments and notifications
for viewers

Commit:
modified:   screenview.cpp
	modified:   screenview.h

FINAL FIX:
adhered to osg::notify mechanism and remove void statements, pointed
out by Jordi.

Commit:
tested and running compilation, ready to be pulled.
comment on the changes: the problem why the viewer didn't work was due to the new 3rdParty library setup,
which now appears to work like a charm. An expected error will be raised ONCE with the OpenSceneGraph head,
as it lacks the following functions:

        inline void set( value_type x, value_type y ) { _v[0]=x; _v[1]=y; }
        inline void set( const Vec2f& rhs) { _v[0]=rhs._v[0]; _v[1]=rhs._v[1]; }
(to be found in <OSG_TOP>/include/Vec2f)

this needs to be patched (currently: manually).

 Changes to be committed:
	modified:   org.openscenegraph.android/jni/Android.mk
	modified:   org.openscenegraph.android/jni/Application.mk
	modified:   org.openscenegraph.osgsimple/src/org/openscenegraph/osgsimple/OSGActivity.java

Commit:
cleared local paths from the Android.mk again

 Changes to be committed:
	modified:   org.openscenegraph.android/jni/Android.mk

jtorresfabra commented:
Hey @CKehl, is this ready for the last review and merge?

CKehl commented:
yes, it is. Just tested it yesterday.
There is one function not available in the main OpenSceneGraph trunk 
concerning the Vec2(f) - it lacks a setter which I needed to insert locally.

Apart from that, it's dead-simple:
Have an OpenSceneGraph trunk (I used 3.3.7 and 3.3.8) ready, compiled 
with plugins for gles1 and gles2 and for armeabi and armeabi-v7a. An 
example script that I used (here GLES1 armeabi):

mkdir build_gles1_obj_armeabi && cd build_gles1_obj_armeabi
cmake .. 
-DCMAKE_TOOLCHAIN_FILE=../PlatformSpecifics/Android/android.toolchain.cmake 
-DDYNAMIC_OPENTHREADS=OFF -DDYNAMIC_OPENSCENEGRAPH=OFF \
-DANDROID_NATIVE_API_LEVEL=8 -DANDROID_ABI=armeabi 
-DCMAKE_INSTALL_PREFIX=${OSGANDROID_TARGET}/gles1/armeabi 
-DOSG_CPP_EXCEPTIONS_AVAILABLE=ON \
-DOSG_GL1_AVAILABLE=OFF -DOSG_GL2_AVAILABLE=OFF -DOSG_GL3_AVAILABLE=OFF 
-DOSG_GLES1_AVAILABLE=ON -DOSG_GLES2_AVAILABLE=OFF 
-DOPENGL_PROFILE="GLES1" \
-DOSG_GL_LIBRARY_STATIC=OFF -DCURL_DIR=../3rdparty/curl 
-DGDAL_DIR=../3rdparty/gdal/include \
-DGIFLIB_INCLUDE_DIR=../3rdparty/giflib 
-DJPEG_INCLUDE_DIR=../3rdparty/libjpeg 
-DPNG_INCLUDE_DIR=../3rdparty/libpng 
-DTIFF_INCLUDE_DIR=../3rdparty/libtiff \
-DZLIB_INCLUDE_DIR=../3rdparty/zlib
make -j 6 install
cd ..
rm -Rf build_gles1_obj_armeabi

then modify the paths according to your installation in the Android.mk 
of org.openscenegraph.android/jni.
ndk-build

And you can test it with the osgsimple application you already provided 
with the whole package.

On 20. okt. 2016 13:31, Jordi Torres wrote:

> Hey @CKehl https://github.com/CKehl, is this ready for the last 
> review and merge?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> https://github.com/miragetech/osgAndroid/pull/2#issuecomment-255080770, 
> or mute the thread 
> https://github.com/notifications/unsubscribe-auth/AFvds4FHy3dGuIzwDiZ4m_rMC4fT5YM8ks5q11ETgaJpZM4Jxzpn.
